vm/ubpf_jit_x86_64.c

```diff
@@ -99,6 +97,410 @@ map_register(int r)
+}
+
+static inline void
+emit_ret(struct jit_state* state)
+{
+    emit1(state, 0xc3);
+}
+
+static inline void
+emit_jmp(struct jit_state* state, uint32_t target_pc)
+{
+    emit1(state, 0xe9);
+    emit_jump_target_address(state, target_pc);
+}
+
+static inline void
+emit_call_through_rax(struct jit_state* state)
+{
+    /*
+     * Caller is responsible for handling stack alignment.
+     */
+#ifndef UBPF_DISABLE_RETPOLINES
+    emit1(state, 0xe8); // e8 is the opcode for a CALL
+    emit_jump_target_address(state, TARGET_PC_RETPOLINE);
+#else
+    /* TODO use direct call when possible */
+    /* callq *%rax */
+    emit1(state, 0xff);
+    // ModR/M byte: b11010000b = xd
+    //               ^
+    //               register-direct addressing.
+    //                 ^
+    //                 opcode extension (2)
+    //                    ^
+    //                    rax is register 0
+    emit1(state, 0xd0);
+#endif
+}
+
+static inline void
+emit_win32_create_home(struct jit_state* state)
+{
+    (void)state;
+#if defined(_WIN32)
+    /* We have to create a little space to keep our 16-byte
+     * alignment happy
+     */
+    emit_alu64_imm32(state, 0x81, 5, RSP, sizeof(uint64_t));
+
+    /* Windows x64 ABI spills 5th parameter to stack */
+    emit_push(state, map_register(5));
+
+    /* Windows x64 ABI requires home register space.
+     * Allocate home register space - 4 registers.
+     */
+    emit_alu64_imm32(state, 0x81, 5, RSP, 4 * sizeof(uint64_t));
+#endif
+}
+
+static inline void
+emit_win32_destroy_home(struct jit_state* state)
+{
+    MAYBE_UNREFERENCED_PARAMETER(state);
+#if defined(_WIN32)
+    /* Deallocate home register space + spilled register + alignment space - 5 registers */
+    emit_alu64_imm32(state, 0x81, 0, RSP, (4 + 1 + 1) * sizeof(uint64_t));
+#endif
+}
+
+static inline void
+emit_call(struct jit_state* state, void* target)
+{
+    /*
+     * When we enter here, our stack is 16-byte aligned. Keep
+     * it that way!
+     */
+    emit_win32_create_home(state);
+
+    emit_load_imm(state, RAX, (uintptr_t)target);
+    emit_call_through_rax(state);
+
+    emit_win32_destroy_home(state);
+}
+
+static inline void
+emit_callx(struct jit_state* state, struct ubpf_vm* vm, int src)
+{
+
+    emit_win32_create_home(state);
+
+    // Because we are going to make a call, we have to preserve
+    // volatile registers. It's great news that the number of
+    // volatile registers maintains 16-byte stack alignment!
+    emit_push(state, RDI);
+    emit_push(state, RSI);
+    emit_push(state, RDX);
+    emit_push(state, RCX);
+    emit_push(state, R8);
+    emit_push(state, R9);
+    emit_push(state, R10);
+    emit_push(state, R11);
+
+    // The first argument (RDI) to ubpf_lookup_registered_function_by_id
+    // is the VM. The second (RSI) is the ID of the helper function to call.
+    emit_load_imm(state, RDI, (uint64_t)vm);
+    emit_mov(state, src, RSI);
+
+    emit_call(state, ubpf_lookup_registered_function_by_id);
```

Comment:
Question: How will this work if the code is moved to a different address space? I.e. JIT in user mode, then execute in kernel mode?